File: KMeans_Clustering.R
# kmeans_clustering.R

# Install ggplot2 if you don't have it
# install.packages("ggplot2")

# Load the ggplot2 library for visualization
library(ggplot2)

# --- 1. Generate Sample Data ---
# Set a seed for reproducibility
set.seed(170)

# Create a dataset with 4 distinct clusters
# We'll generate data from 4 different normal distributions
c1 <- data.frame(x = rnorm(75, 1, 1), y = rnorm(75, 8, 1), cluster = "1")
c2 <- data.frame(x = rnorm(75, 8, 1), y = rnorm(75, 2, 1), cluster = "2")
c3 <- data.frame(x = rnorm(75, -5, 1), y = rnorm(75, 2, 1), cluster = "3")
c4 <- data.frame(x = rnorm(75, -2, 1), y = rnorm(75, -5, 1), cluster = "4")

# Combine into a single dataframe
data <- rbind(c1, c2, c3, c4)

# We will only use the feature columns for clustering
data_for_clustering <- data[, c("x", "y")]


# --- 2. Apply the K-Means Algorithm ---
# We use the built-in kmeans() function from the 'stats' package (loaded by default)
# We set centers=4 because we know there are 4 groups.
# nstart is the number of random sets to choose.
kmeans_result <- kmeans(data_for_clustering, centers = 4, nstart = 25)

# Add the cluster assignments back to the original data frame
data$assigned_cluster <- as.factor(kmeans_result$cluster)

# --- 3. Visualize the Results ---
# Create a scatter plot of the data points, colored by their assigned cluster
plot <- ggplot(data, aes(x = x, y = y, color = assigned_cluster)) +
  geom_point(size = 3) +
  labs(
    title = "K-Means Clustering in R",
    x = "Feature 1",
    y = "Feature 2",
    color = "Cluster"
  ) +
  theme_minimal() +
  theme(legend.position = "right")

# Add the cluster centers to the plot
centers <- as.data.frame(kmeans_result$centers)
plot <- plot + geom_point(data = centers, aes(x = x, y = y), color = "red", size = 5, shape = 4, stroke = 2)

# Save the plot to a file
ggsave("kmeans_clusters_r.png", plot)

print("K-Means clustering complete. The visualization has been saved to 'kmeans_clusters_r.png'")

# Show the plot
print(plot)

File: kmeans_clustering.py
# kmeans_clustering.py

# Import necessary libraries
import matplotlib.pyplot as plt
from sklearn.cluster import KMeans
from sklearn.datasets import make_blobs

# --- 1. Generate Sample Data ---
# We will generate a synthetic dataset with 4 distinct clusters.
n_samples = 300
random_state = 170
X, y = make_blobs(n_samples=n_samples, random_state=random_state, centers=4)

# --- 2. Apply the K-Means Algorithm ---
# We initialize KMeans with n_clusters=4, as we know there are 4 blobs.
# 'n_init' is set to 'auto' to let scikit-learn determine the best number of initializations.
# 'random_state' is set for reproducibility of results.
kmeans = KMeans(n_clusters=4, n_init='auto', random_state=random_state)

# Fit the model to the data and predict the cluster for each data point.
y_pred = kmeans.fit_predict(X)

# --- 3. Visualize the Results ---
# Create a scatter plot of the data points, colored by their assigned cluster.
plt.figure(figsize=(8, 6))
plt.scatter(X[:, 0], X[:, 1], c=y_pred, s=50, cmap='viridis')

# Plot the cluster centers (centroids) as red 'X's.
centers = kmeans.cluster_centers_
plt.scatter(centers[:, 0], centers[:, 1], c='red', s=200, alpha=0.75, marker='X')

# Add titles and labels for clarity.
plt.title('K-Means Clustering')
plt.xlabel('Feature 1')
plt.ylabel('Feature 2')
plt.grid(True)

# Save the plot to a file
plt.savefig('kmeans_clusters_python.png')

print("K-Means clustering complete. The visualization has been saved to 'kmeans_clusters_python.png'")

# Show the plot
plt.show()
